### DISCUSSION TITLE
magit-process: Second-level collapsible sections from hook outputs

### DISCUSSION THREAD

**Discussion by MicahElliott** [magit/magit#5030] [Ideas]
I have several git-hooks that show (some) important output. (It happens to be from a git-hook manager which I think commonly could have lots of output.) Some of the hooks show a little too much output (though it's still sometimes useful for diagnoses), so I'm wondering if it would be possible to make those long outputs sectioned/collapsed/collapsible in the same way that each git command output is already done in magit-process. Eg,

```
  0 git … commit --…

    git … commit --
foolinter running
  ▶ Looking at a bunch of files:… (collapsible/expandable subsection because of leading right-triangle, maybe?)
```

There is a lot of auto-collapsing sprinkled around magit, so I'm not sure if there's already a simple way (some syntax for before the output?) to just print something so that lines following it (maybe at specific spaced indentation?) can be a subsection.

I'd guess these could be hierarchically navigable in the same way that `magit-status` sections are with `n`, `p`, `M-n`, `M-p`, `^`, `TAB`, etc.

I'm throwing this out there in case you think it's a good idea, or if there's some simple means to achieve it.

**Comment by tarsius** [comment]
Adding this would have a performance impact and it also wouldn't be easy to implement. How exactly to split the output into sections very much depends on the used git hook. So this would have to be done in a generic fashion that third-party code could hook into. Even if that were not the case, it still would be a major case. Currently we just insert the output into "a command's" section in the process buffer and we are done with it. Either parsing the output as it arrives or else continuing to insert as-is, but then going back to transform it once the process has finished, are both expensive and error-prone.

**Comment by MicahElliott** [comment]
Thanks Jonas. That makes sense. Appreciate the detailed explanation.

I did find a workaround in case anyone else wants this...

- Install [page-break-lines package](https://github.com/purcell/page-break-lines)
- Have your git hooks separate their output with page breaks (`^L`)
- Turn on `page-break-lines-mode` for buffers, particularly adding `magit-process-mode` to `page-break-lines-modes`

Now you can navigate around sections with `C-x [` and `C-x ]`, and the page breaks will have a nice colored line styled to your liking.

Example in screenshot.

![2023-10-16-1029_scrot](https://github.com/magit/magit/assets/159047/2e8a370d-2227-4a14-bdaa-48d1f3c040db)
